### PR TITLE
Feat/soap

### PR DESCRIPTION
**Pull request by felixmusil**
Implementation of the soap representation in 2 modes. 

Tests and bindings are there too.

### CODE REVIEW

File: bindings/bind_py_feature_manager.cc [hunk 1]
```diff
@@ -29,41 +29,82 @@
 
 using namespace rascal;  // NOLINT
 
+template <typename T>
+void bind_feature_manager_base(py::module & m_garbage) {
+  using Base_t = FeatureManagerBase<T>;
+
+  std::string featurebase_name = internal::GetBindingTypeName<Base_t>();
+  py::class_<Base_t>(m_garbage, featurebase_name.c_str());
```
Comment (max-veit) [inline bindings/bind_py_feature_manager.cc]: I still don't like the name `m_garbage` (could be confused with https://docs.python.org/3/library/gc.html) - try `m_throwaway` or `m_placeholder`?

File: scripts/soap_example.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    return y
+
+##########################################################################################
+
+def normalise(spectrum):
+    x = spectrum
+    ncen = spectrum.shape[0]
+    for i in range(ncen):
+        norm = np.linalg.norm(x[i])
+        if norm >= 1.0e-20: x[i] /= norm
```
Comment (max-veit) [inline scripts/soap_example.py]: Is this supposed to be some sort of `eps`-related value? Either way, it should be defined as a constant somewhere.

File: src/math/math_utils.cc [hunk 3]
```diff
@@ -228,6 +227,7 @@ namespace rascal {
           }  // if (m_count == 0)
         }    // for (m_count in [0, l])
       }      // for (l in [0, lmax])
+      // harmonics *= std::sqrt(2.0);
```
Comment (max-veit) [inline src/math/math_utils.cc]: Why is this here? As a comment, it's extremely confusing - if it works without it, then just delete this line.

File: src/representations/feature_manager_dense.hh [hunk 4]
```diff
@@ -159,6 +159,10 @@ namespace rascal {
                                 this->feature_size(), this->sample_size());
     }
 
+    inline Feature_Matrix_t get_feature_matrix_dense() {
+      return Feature_Matrix_t::Zero(2, 2);
```
Comment (max-veit) [inline src/representations/feature_manager_dense.hh]: ...Is this a placeholder? Otherwise, I don't understand what this is supposed to do.

File: src/structure_managers/structure_manager.hh [hunk 5]
```diff
@@ -809,6 +809,10 @@ namespace rascal {
   std::array<int, Order>
   StructureManager<ManagerImplementation>::ClusterRef<Order>::get_atom_types()
       const {
+    /*
```
Comment (max-veit) [inline src/structure_managers/structure_manager.hh]: ...

File: tests/test_representation_manager.cc [hunk 6]
```diff
@@ -178,87 +179,170 @@ namespace rascal {
...
           }
         }
       }
+      manager_i += 1;
     }
   }
 
   BOOST_AUTO_TEST_SUITE_END();
 
+  // /* ----------------------------------------------------------------------
```
Comment (max-veit) [inline tests/test_representation_manager.cc]: What's the purpose of all this commented stuff?

File: src/representations/representation_manager_sorted_coulomb.hh [hunk 7]
```diff
@@ -65,13 +65,14 @@ namespace rascal {
     // Enum class defining the several possible sorting options of the Coulomb
     // Matrix
     enum class CMSortAlgorithm {
-      Distance, // sort according to the distance from the central atom
-      RowNorm, // sort according to the norm of the coulomb matrix rows
+      Distance,  // sort according to the distance from the central atom
```
Comment (max-veit) [inline src/representations/representation_manager_sorted_coulomb.hh]: Ideally there shouldn't be any cosmetic changes to unrelated files (sorted-coulomb, in a soap pull request)

File: src/math/math_utils.cc (hunk 3 above)
Comment (max-veit) [reply src/math/math_utils.cc]: OK, I think it's left over from testing, so definitely delete it.

Comment (max-veit) [reply src/math/math_utils.cc]:
```suggestion
```

File: src/representations/feature_manager_dense.hh (hunk 4 above)
Comment (felixmusil) [reply src/representations/feature_manager_dense.hh]: yes it is. I added a static_assert to avoid any call to it. In the near future the feature manager will disappear anyway.

File: tests/test_representation_manager.cc (hunk 6 above)
Comment (felixmusil) [reply tests/test_representation_manager.cc]: forgot to delete it.

File: src/representations/representation_manager_sorted_coulomb.hh (hunk 7 above)
Comment (felixmusil) [reply src/representations/representation_manager_sorted_coulomb.hh]: well this is automatic

**Comment by max-veit** [comment]
@mastricker anything to add before we merge?

File: src/representations/representation_manager_spherical_expansion.hh [hunk 8]
```diff
@@ -419,68 +437,81 @@ namespace rascal {
...
     for (auto center : this->structure_manager) {
-      Eigen::MatrixXd soap_vector = Eigen::MatrixXd::Zero(
-          this->n_species * this->max_radial, pow(this->max_angular + 1, 2));
-      Eigen::MatrixXd radial_integral(this->max_radial, this->max_angular + 1);
-      // Eigen::MatrixXd radial_integral =
-      //     Eigen::MatrixXd::Zero(this->max_radial, this->max_angular + 1);
+      auto & coefficients_center = this->expansions_coefficients[center];
+
+      // initialize the expansion coefficients to 0
+      key_t center_type{center.get_atom_type()};
```
Comment (mastricker) [inline src/representations/representation_manager_spherical_expansion.hh]: `Key_t`

File: src/representations/representation_manager_spherical_expansion.hh [hunk 9]
```diff
@@ -419,68 +437,81 @@ namespace rascal {
...
+      coefficients_center[center_type] = dense_t::Zero(n_row, n_col);
+      for (auto neigh : center) {
+        key_t neigh_type{neigh.get_atom_type()};
```
Comment (mastricker) [inline src/representations/representation_manager_spherical_expansion.hh]: `Key_t`

File: src/structure_managers/property_base.hh [hunk 10]
```diff
@@ -91,6 +92,12 @@ namespace rascal {
       this->update_nb_comp();
     }
 
+    inline void set_shape(const Dim_t & nb_row, const Dim_t & nb_col) {
```
Comment (mastricker) [inline src/structure_managers/property_base.hh]: I guess this should serve as a shorthand instead of using `set_nb_row()` and `set_nb_col()`?
Makes sense.

File: src/structure_managers/structure_manager.hh [hunk 11]
```diff
@@ -809,6 +809,10 @@ namespace rascal {
...
+     * TODO(markus)
```
Comment (mastricker) [inline src/structure_managers/structure_manager.hh]: I am not sure what this todo is referring to?
Could you please elaborate?

File: src/structure_managers/property_block_sparse.hh [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+namespace rascal {
+
+  namespace internal {
+
+    template <class K, class V>
+    class InternallySortedKeyMap {
+     public:
+      using MyMap_t = std::map<K, V>;
+
+      //! the data holder. only the overiden/essential functionalities are
```
Comment (mastricker) [inline src/structure_managers/property_block_sparse.hh]: *overriden

File: src/structure_managers/property_block_sparse.hh [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+
+      template <typename... Args>
+      decltype(auto) emplace(Args &&... args) {
+        return this->data.emplace(std::forward<Args>(args)...);
+      }
+
+      /**
+       * returns a vector of the valid keys of the map
+       */
+      std::vector<key_type> get_keys() {
```
Comment (mastricker) [inline src/structure_managers/property_block_sparse.hh]: I only stumbled over it here - What can `key_type` all mean? It is invoked with a class template `K` - is this intentional?

File: src/structure_managers/property_block_sparse.hh [hunk 14]
```diff
@@ -0,0 +1,600 @@
...
+    template <class K, class V>
+    class InternallySortedKeyMap {
+     public:
+      using MyMap_t = std::map<K, V>;
+
+      //! the data holder. only the overiden/essential functionalities are
+      //! directly exposed
+      MyMap_t data{};
+
+      // some member types
```
Comment (mastricker) [inline src/structure_managers/property_block_sparse.hh]: Member types in camelcase or is this for compliance with std library?

File: src/structure_managers/property_block_sparse.hh [hunk 15]
```diff
@@ -0,0 +1,600 @@
...
+  }  // namespace internal
+  /* ---------------------------------------------------------------------- */
+  /**
+   * Typed ``property`` class definition, inherits from the base property class
+   */
+  template <typename Precision_t, size_t Order, size_t PropertyLayer>
+  class BlockSparseProperty : public PropertyBase {
+   public:
+    using Parent = PropertyBase;
+    using dense_t = Eigen::Matrix<Precision_t, Eigen::Dynamic, Eigen::Dynamic>;
```
Comment (mastricker) [inline src/structure_managers/property_block_sparse.hh]: types in camelcase

File: src/structure_managers/property_block_sparse.hh [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+      return this->values[index];
+    }
+
+    inline size_t get_dense_feature_size(const size_t & index) {
+      auto keys = this->values[index].get_keys();
+      return this->get_nb_comp() * keys.size();
+    }
+
+    //! Accessor for property by cluster index and return a dense
+    //! representation of the property associated to this cluster
```
Comment (mastricker) [inline src/structure_managers/property_block_sparse.hh]: Representation is the right word in this context, but given our use or representation, we should probably use something else? Or do you think this is unambiguous?

File: src/structure_managers/property_block_sparse.hh [hunk 17]
```diff
@@ -0,0 +1,600 @@
...
+        const auto & value{element.second};
+        if (value.size() != this->get_nb_comp()) {
+          auto error{std::string("Size should match: ") +
+                     std::to_string(value.size()) + std::string(" != ") +
+                     std::to_string(this->get_nb_comp())};
+          throw std::length_error(error);
+        }
+      }
+
+      this->values.push_back(ref);
```
Comment (mastricker) [inline src/structure_managers/property_block_sparse.hh]: in this vector `values`, which is a vector of the internal key map: I am a little confused. When it is a map, how can it be access later with `[]` with and index in `operator[]`?

File: src/structure_managers/property_block_sparse.hh [hunk 18]
```diff
@@ -0,0 +1,600 @@
...
+    data_t values{};  //!< storage for properties
+    sizes_t center_sizes{};
+    keys_t all_keys{};
+    keys_list_t keys_list{};
+  };
+
+  // /* ----------------------------------------------------------------------
+  // */
+  // /**
+  //  * Typed ``property`` class definition, inherits from the base property
```
Comment (mastricker) [inline src/structure_managers/property_block_sparse.hh]: is this part necessary?

File: tests/test_feature_manager.cc [hunk 19]
```diff
@@ -123,6 +117,126 @@ namespace rascal {
...
+                i_count++;
+              }
+              if (verbose)
+                std::cout << std::endl;
+            }
+          } else {
+            i_count += inner_size;
+          }
+        }
+        BOOST_CHECK_LE(diff, 1e-11);
```
Comment (mastricker) [inline tests/test_feature_manager.cc]: Is it possible to add to this test now only columns/rows but access with clusters?
This would show that the map works?
Or am I missing something?

File: tests/test_properties.cc [hunk 20]
```diff
@@ -205,4 +207,134 @@ namespace rascal {
 
   BOOST_AUTO_TEST_SUITE_END();
 
+  /* ---------------------------------------------------------------------- */
+
+  /*
+   * A fixture for testing partially sparse proterties.
```
Comment (mastricker) [inline tests/test_properties.cc]: *properties

File: tests/test_properties.cc [hunk 21]
```diff
@@ -205,4 +207,134 @@ namespace rascal {
...
+   * TODO(felix) use MultipleStructureManagerCentersFixture instead of NL
+   */
+  struct BlockSparsePropertyFixture
+      : public MultipleStructureFixture<MultipleStructureManagerNLFixture> {
+    using Parent = MultipleStructureFixture<MultipleStructureManagerNLFixture>;
+    using ManagerTypeList_t = typename Parent::ManagerTypeHolder_t::type_list;
+
+    using key_t = std::vector<int>;
```
Comment (mastricker) [inline tests/test_properties.cc]: camelcase for types

**Comment by felixmusil** [comment]
Thanks for your comments.

1. it is missing but it is not too clear to me what should be tested that is not already covered in the cpp tests except checking the imports are working and that is runs
2. I improved the notebook and renamed it
3. It depends on the spherical expansion which is what we want since the spherical expansion is an ingredient in other representations.
4. the dummy is there for convenience and should disappear after the structure manager aggregator
5. Let's discuss more about the aggregator next meeting
6. it is defined in 2 different test suite so no problems. 
7. it could be split but we should be able to merge these later on so I would leave it at that for the moment

File: examples/test_nl.cc [hunk 22]
```diff
@@ -53,155 +54,98 @@ using Vector_t = Eigen::Matrix<double, dim, 1>;
...
-
-  for (auto && center : my_man) {
-    if (verbose) {
-      std::cout << "################################# 2" << std::endl;
-      std::cout << center.get_atom_type() << std::endl;
+  using ManagerTypeList_t = typename ManagerTypeHolder_t::type_list;
+  using Manager_t = typename ManagerTypeHolder_t::type;
+  using Std2DArray_t = std::vector<std::vector<double>>;
+
+  auto dd{TestData()};
```
[the remark this answers is not in the document]
Comment (felixmusil) [reply examples/test_nl.cc]: corrected

File: scripts/soap_example.py [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+
+##########################################################################################
+
+#dump radial and power spectra for methane
+def dump_reference_json():
+    import ubjson
+    import os
+    from copy import copy
+    path = '../'
+    sys.path.insert(0, os.path.join(path, 'build/'))
```
[the remark this answers is not in the document]
Comment (felixmusil) [reply scripts/soap_example.py]: it is necessary

[the remark this answers is not in the document]
Comment (mastricker) [reply scripts/soap_example.py]: I can comment it out and it still works?
There is an import on line 3 for the whole file?

File: scripts/soap_example.py [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+
+#------------------------------------------nu=2------------------------------------------#
+
+    test_hypers["soap_type"] = "PowerSpectrum"
+    x = get_spectrum(test_hypers, frames)
+    x = x.T #Eigen column major
+    x = unravel_power_spectrum(x, nspecies, ncen)
+    x = normalise(x)
+    kernel = np.dot(x, x.T)
+    np.save('kernel_soap_example_nu2.npy', kernel)
```
[the remark this answers is not in the document]
Comment (felixmusil) [reply scripts/soap_example.py]: we stick with pyhton 3 so there is not problem with pickle compatibility and the danger of this format is not an issue for this use case in my opinion

**Comment by mastricker** [comment]
Hi,

1. I am not saying to test everything again on the python side, but where is the part "that it runs"?
What I meant is: test which show that the bindings actually work. Input to python functions, evaluation "compute" and see if an expected value for whatever comes out?
2. Ok, that's good. But the notebook now includes output of the cells, including a pictures in textform...
We should probably also add a *npy in the `.gitignore` for the output 
3. Ok, let's discuss that also. I am not sure how that will be handled with the `RepresentationCalculator`. I put it in the meeting notes.
4. Ok, I see.
5. Ok, I put up a point for the discussion.
6. Ok.
7. Ok, but just to make my point: I really want to split that up later. Can we make a todo comment in the file so that we stumble over it?

So far so good - what about the other comments related to code? Agree/Disagree?

File: examples/test_nl.cc [hunk 25]
```diff
@@ -48,104 +47,34 @@
...
+  json ad2{{"name", "AdaptorStrict"},
+           {"initialization_arguments", {{"cutoff", cutoff}}}};
+  adaptors.emplace_back(ad1);
+  adaptors.emplace_back(ad2);
+  auto manager =
+      make_structure_manager_stack<StructureManagerCenters,
+                                   AdaptorNeighbourList, AdaptorStrict>(
+          structure, adaptors);
+  Representation_t representation{manager, hypers};
+  representation.compute();
```
[the remark this answers is not in the document]
Comment (felixmusil) [reply examples/test_nl.cc]: this is not a test but an example :)

File: src/structure_managers/structure_manager.hh (hunk 11 above)
Comment (felixmusil) [reply src/structure_managers/structure_manager.hh]: I probably meant to document a bit this part on the functionality

File: src/structure_managers/property_block_sparse.hh (hunk 14 above)
Comment (felixmusil) [reply src/structure_managers/property_block_sparse.hh]: yes

File: tests/python/python_binding_tests.py [hunk 26]
```diff
@@ -1,4 +1,3 @@
-#!/usr/bin/env python3
```
Comment (max-veit) [inline tests/python/python_binding_tests.py]: This is the change that broke the tests. Please make sure your editor doesn't accidentally delete the shebang line!
